File: src/main/snowflake_quote.py
# snowflake_quote
# A Python3 function to properly escape and quote string literals for Snowflake
#
# Copyright (c) 2022 Ronald B. Cemer
# Licensed under the MIT license.
#
# Project: https://github.com/roncemer/snowflake_quote

def snowflake_quote(s):
    if s is None:
        return "NULL"

    result = "'"
    for c in s:
        if (c == "'" or c == '"' or c == '\\'):
            result = result + '\\' + c
        elif (c == '\b'):
            result = result + '\\b'
        elif (c == '\f'):
            result = result + '\\f'
        elif (c == '\n'):
            result = result + '\\n'
        elif (c == '\r'):
            result = result + '\\r'
        elif (c == '\t'):
            result = result + '\\t'
        else:
            cc = ord(c)
            if ((cc >= 0x00 and cc <= 0x1f) or cc >= 0x7f):
                result = result + ("\\x%0.2x" % cc)
            else:
                result = result + c
    result = result + "'"
    return result

File: src/main/snowflake_quote.js
module.exports.snowflake_quote = function(s) {
    if (s === null) return 'NULL';

    let result = "'"
    for (let i = 0, n = s.length; i < n; i++) {
        let c = s[i];
        if (c == "'" || c == '"' | c == '\\') {
            result = result + '\\' + c;
        } else if (c == '\b') {
            result = result + '\\b';
        } else if (c == '\f') {
            result = result + '\\f';
        } else if (c == '\n') {
            result = result + '\\n';
        } else if (c == '\r') {
            result = result + '\\r';
        } else if (c == '\t') {
            result = result + '\\t';
        } else {
            let cc = c.charCodeAt(0);
            if ((cc >= 0x00 && cc <= 0x1f) || cc >= 0x7f) {
                cc = cc.toString(16);
                if (cc.length < 2) cc = "0" + cc;
                result = result + "\\x" + cc;
            } else {
                result = result + c;
            }
        }
    }
    result = result + "'"
    return result
}
